wallet: cache IsMine scriptPubKeys to improve performance of descriptor wallets

achow101 opened the pull request:
Wallets that have a ton of non-ranged descriptors (such as a migrated non-HD wallet) perform fairly poorly due to looping through all of the wallet's `ScriptPubKeyMan`s. This is done in various places, such as `IsMine`, and helper functions for fetching a `ScriptPubKeyMan` and a `SolvingProvider`. This also has a bit of a performance impact on standard descriptor wallets, although less noticeable due to the small number of SPKMs.

As these functions are based on doing `IsMine` for each `ScriptPubKeyMan`, we can improve this performance by caching `IsMine` scriptPubKeys for all descriptors and use that to determine which `ScriptPubKeyMan` to actually use for those things. This cache is used exclusively and we no longer iterate the SPKMs.

Also added a benchmark for `IsMine`.

achow101 commented:
Dropped the signing changes as they were causing issues with being able to sign transactions that we have the keys for but not the descriptors. #23417 would allow us to resolve the performance issues for signing while retaining this behavior.

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -3425,13 +3457,24 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
+
+    // Cache scriptPubKeys if the script is not ranged
+    ScriptPubKeyMan* spkm = m_spk_managers.at(id).get();
+    m_uncached_spkms.insert(spkm);
+    if (!ranged) {
+        for (const auto& script : spkm->GetScriptPubKeys()) {
+            m_cached_spks[script].insert(spkm);
+            m_uncached_spkms.erase(spkm);
+        }
+    }
```
```suggestion
    if (!ranged) {
        for (const auto& script : spkm->GetScriptPubKeys()) {
            m_cached_spks[script].insert(spkm);
        }
    }
    else {
        m_uncached_spkms.insert(spkm);
    }
```

achow101 replied:
I prefer the way it was written as it ensures that no spkm will be accidentally missed.

w0xlt commented on src/bench/wallet_balance.cpp:
```diff
@@ -31,7 +30,6 @@ static void WalletBalance(benchmark::Bench& bench, const bool set_dirty, const b
         LOCK(wallet.cs_wallet);
         wallet.SetWalletFlag(WALLET_FLAG_DESCRIPTORS);
         wallet.SetupDescriptorScriptPubKeyMans();
-        if (wallet.LoadWallet() != DBErrors::LOAD_OK) assert(false);
```
What would be the reason to remove this validation ?

w0xlt commented on src/wallet/wallet.h:
```diff
@@ -341,6 +341,11 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     std::map<uint256, std::unique_ptr<ScriptPubKeyMan>> m_spk_managers;
 
+    //! Cache of descriptor ScriptPubKeys used for IsMine. Maps ScriptPubKey to set of spkms
+    std::unordered_map<CScript, std::unordered_set<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks;
+    //! Set of which descriptors are not in m_cached_spks
+    std::unordered_set<ScriptPubKeyMan*> m_uncached_spkms;
```
I think the names below work best and make the purpose of the variable clearer.

```suggestion
    std::unordered_map<CScript, std::unordered_set<ScriptPubKeyMan*>, SaltedSipHasher> m_ranged_spks;
    //! Set of which descriptors are not in m_cached_spks
    std::unordered_set<ScriptPubKeyMan*> m_non_ranged_spkms;
```

achow101 replied:
I prefer describing the variable with how we expect to use it, rather than what it contains. We're using it as a cache of spkms, the type of spkm doesn't matter.

w0xlt reviewed (commented):
This change looks good, but adds complexity by adding two caches instead of accessing `m_spk_managers` directly.

Perhaps the PR description might include benchmark improvement as a reason for the added complexity.

aureleoules reviewed (commented):
Concept ACK

achow101 commented on src/bench/wallet_ismine.cpp:
```diff
@@ -0,0 +1,105 @@
+
+static const std::shared_ptr<CWallet> BenchLoadWallet(std::unique_ptr<WalletDatabase> database, WalletContext& context, DatabaseOptions& options)
+{
+    bilingual_str error;
+    std::vector<bilingual_str> warnings;
+    auto wallet = CWallet::Create(context, "", std::move(database), options.create_flags, error, warnings);
+    NotifyWalletLoaded(context, wallet);
+    if (context.chain) {
+        wallet->postInitProcess();
+    }
+    return wallet;
+}
+
+static void BenchUnloadWallet(std::shared_ptr<CWallet>&& wallet)
+{
+    SyncWithValidationInterfaceQueue();
+    wallet->m_chain_notifications_handler.reset();
+    UnloadWallet(std::move(wallet));
+}
+
+static void WalletIsMine(benchmark::Bench& bench, bool legacy_wallet, int num_combo = 0)
+{
+    const auto test_setup = MakeNoLogFileContext<TestingSetup>();
+    test_setup->m_args.ForceSetArg("-unsafesqlitesync", "1");
+
+    WalletContext context;
+    context.args = &test_setup->m_args;
+    context.chain = test_setup->m_node.chain.get();
+
+    // Setup the wallet
+    // Loading the wallet will also create it
+    DatabaseOptions options;
+    if (legacy_wallet) {
+        options.require_format = DatabaseFormat::BERKELEY;
+    } else {
+        options.create_flags = WALLET_FLAG_DESCRIPTORS;
+        options.require_format = DatabaseFormat::SQLITE;
+    }
+    auto database = CreateMockWalletDatabase(options);
+    auto wallet = BenchLoadWallet(std::move(database), context, options);
```
Refactored.

darosior commented:
Concept ACK. This was hit in practice by a user in #27002. From https://github.com/bitcoin/bitcoin/issues/27002#issuecomment-1412556699:
- v24.0.1:
    ```
    $ time ./bitcoin-cli -rpcwallet=desc listunspent|grep txid|wc -l
    224
    
    real	1m15.596s
    user	0m0.008s
    sys	0m0.000s
    ```
- This PR:
    ```
    $ time ./bitcoin-cli -rpcwallet=desc listunspent|grep txid|wc -l
    314

    real	0m0.464s
    user	0m0.009s
    sys	0m0.000s
    ```

willcl-ark commented:
Concept ACK.

I wonder if the performance boost here might also be enough to close out https://github.com/bitcoin/bitcoin/issues/19942 even though it's not implementing the functionality OP there thought would solve their issue?

achow101 commented:
I've ended up rewriting a significant chunk of this PR. Instead of caching the scriptPubKeys of non-ranged descriptors, I've changed this to cache the scriptPubKeys of all descriptors. I think this makes it easier to understand and is also less fragile.

Originally I didn't cache the spks of ranged descriptors because I thought it would be difficult to update the cached spks every time a `TopUp` was called as that is used in several places internally. My solution for that issue is to have a callback in the SPKM that is called by `TopUp` to inform the containing `CWallet` of the newly added spks. This allows the spk cache to always match the descriptor's spk set.

One concern of this approach is that it maintains duplicate copies of the scriptPubKeys, thus significantly increasing the memory footprint. However, this may not really be a concern given that scriptPubKeys are relatively small and so it really just ends up being a couple of kB.

This also managed to improve the performance of `IsMine` even without a ton of non-ranged descriptors. Here are the benchmark results for the original, and the new:

Original:
|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|           11,445.75 |           87,368.70 |    0.2% |      103,638.12 |      18,444.02 |    0.1% |      0.01 | `WalletIsMineDescriptors`
|            8,256.94 |          121,110.23 |    0.1% |       74,299.09 |      10,755.01 |    0.1% |      0.01 | `WalletIsMineLegacy`
|           12,289.10 |           81,372.91 |    0.3% |      114,209.13 |      20,204.02 |    0.1% |      0.01 | `WalletIsMineMigratedDescriptors`

New:
|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|            1,734.82 |          576,429.48 |    0.2% |       15,191.02 |       2,579.00 |    0.0% |      0.01 | `WalletIsMineDescriptors`
|            8,372.88 |          119,433.26 |    0.2% |       77,389.09 |      11,220.01 |    0.1% |      0.01 | `WalletIsMineLegacy`
|            1,753.04 |          570,438.89 |    0.4% |       15,359.02 |       2,606.00 |    0.0% |      0.01 | `WalletIsMineMigratedDescriptors`

darosior reviewed (commented):
I really like the performance gains here, however i feel like this is making the spks management bleed into the wallet from the spkman. But there is no other way for the wallet to efficiently locate the spkman(s)..

I was looking into how to get rid of the callbacks and avoid duplicating the scriptpubkeys, and thought we might as well share a single mapping owned by `CWallet` across all spkmans. This would also have the nice property that we don't need to make sure moving forward we don't introduce any inconsistency between the `CWallet` cache and what the spkmans track (as we'd have to with the approach here).
However i can't say i like this approach better.. Thought i'd share it here anyway: https://github.com/darosior/bitcoin/tree/pr_26008 (quick and dirty, there is a bug [0], but should be readable enough to have an idea of what this approach could look like). Performances according to the benchmarks introduced here are similar.

---

[0] Interestingly all tests pass, this bug only shows up in one of the benchmark, but not one of those introduced here.

achow101 commented:
> I was looking into how to get rid of the callbacks and avoid duplicating the scriptpubkeys

I added the callback to also allow for other cases where we need to know that the scriptPubKey set has changed, such as with fast rescans with block filters.

achow101 commented:
First 2 commits split into #27666. Marking as draft for now.

aureleoules commented:
The benchmark `WalletCreateTxUsePresetInputsAndCoinSelection` seems to segfault occasionally on this pull (increasing min-time makes it crash consistently)
```bash
$ valgrind ./src/bench/bench_bitcoin -filter=WalletCreateTxUsePresetInputsAndCoinSelection -min-time=10000
==131665== Memcheck, a memory error detector
==131665== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==131665== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==131665== Command: ./src/bench/bench_bitcoin -filter=WalletCreateTxUsePresetInputsAndCoinSelection -min-time=10000
==131665==
==131665== Invalid read of size 8
==131665==    at 0x607B83: wallet::CWallet::AddToWalletIfInvolvingMe(std::shared_ptr<CTransaction const> const&, std::variant<wallet::TxStateConfirmed, wallet::TxStateInMempool, wallet::TxStateInactive> const&, bool, bool) (wallet.cpp:1252)==131665==    by 0x60879A: SyncTransaction (wallet.cpp:1409)
==131665==    by 0x60879A: wallet::CWallet::blockConnected(interfaces::BlockInfo const&) (wallet.cpp:1479)
==131665==    by 0x28B59C: generateFakeBlock(CChainParams const&, node::NodeContext const&, wallet::CWallet&, CScript const&) (wallet_create_tx.cpp:73)
==131665==    by 0x28C982: WalletCreateTx(ankerl::nanobench::Bench&, OutputType, bool, std::optional<PreSelectInputs>) [clone .constprop.0] [clone .isra.0] (wallet_create_tx.cpp:103)
==131665==    by 0x215E9A: operator() (std_function.h:590)
==131665==    by 0x215E9A: benchmark::BenchRunner::RunAll(benchmark::Args const&) (bench.cpp:119)
==131665==    by 0x1F3CB6: main (bench_bitcoin.cpp:132)
==131665==  Address 0x974ad60 is 0 bytes inside a block of size 632 free'd
==131665==    at 0x484BB6F: operator delete(void*, unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==131665==    by 0x5FE039: operator() (unique_ptr.h:85)
==131665==    by 0x5FE039: reset (unique_ptr.h:182)
==131665==    by 0x5FE039: operator= (unique_ptr.h:167)
==131665==    by 0x5FE039: operator= (unique_ptr.h:212)
==131665==    by 0x5FE039: operator= (unique_ptr.h:371)
==131665==    by 0x5FE039: wallet::CWallet::AddScriptPubKeyMan(uint256 const&, std::unique_ptr<wallet::ScriptPubKeyMan, std::default_delete<wallet::ScriptPubKeyMan> >) (wallet.cpp:3509)
==131665==    by 0x615D95: wallet::CWallet::LoadDescriptorScriptPubKeyMan(uint256, wallet::WalletDescriptor&) (wallet.cpp:3556)
==131665==    by 0x63D847: wallet::LoadDescriptorWalletRecords(wallet::CWallet*, wallet::DatabaseBatch&, int)::{lambda(wallet::CWallet*, DataStream&, CDataStream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)#1}::operator()(wallet::CWallet*, DataStream&, CDataStream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const (walletdb.cpp:807)
==131665==    by 0x631799: operator() (std_function.h:590)
==131665==    by 0x631799: wallet::LoadRecords(wallet::CWallet*, wallet::DatabaseBatch&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DataStream&, std::function<wallet::DBErrors (wallet::CWallet*, DataStream&, CDataStream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)>) (walletdb.cpp:509)
==131665==    by 0x631F16: wallet::LoadRecords(wallet::CWallet*, wallet::DatabaseBatch&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<wallet::DBErrors (wallet::CWallet*, DataStream&, CDataStream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)>) (walletdb.cpp:523)
==131665==    by 0x63E81D: LoadDescriptorWalletRecords (walletdb.cpp:789)
==131665==    by 0x63E81D: wallet::WalletBatch::LoadWallet(wallet::CWallet*) (walletdb.cpp:1177)
==131665==    by 0x5F85DD: wallet::CWallet::LoadWallet() (wallet.cpp:2311)
==131665==    by 0x28C913: WalletCreateTx(ankerl::nanobench::Bench&, OutputType, bool, std::optional<PreSelectInputs>) [clone .constprop.0] [clone .isra.0] (wallet_create_tx.cpp:93)
==131665==    by 0x215E9A: operator() (std_function.h:590)
==131665==    by 0x215E9A: benchmark::BenchRunner::RunAll(benchmark::Args const&) (bench.cpp:119)
==131665==    by 0x1F3CB6: main (bench_bitcoin.cpp:132)
==131665==  Block was alloc'd at
==131665==    at 0x4849013: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==131665==    by 0x6011D7: wallet::CWallet::SetupDescriptorScriptPubKeyMans(CExtKey const&) (wallet.cpp:3572)
==131665==    by 0x60B3E9: wallet::CWallet::SetupDescriptorScriptPubKeyMans() (wallet.cpp:3604)
==131665==    by 0x28C90B: WalletCreateTx(ankerl::nanobench::Bench&, OutputType, bool, std::optional<PreSelectInputs>) [clone .constprop.0] [clone .isra.0] (wallet_create_tx.cpp:92)
==131665==    by 0x215E9A: operator() (std_function.h:590)
==131665==    by 0x215E9A: benchmark::BenchRunner::RunAll(benchmark::Args const&) (bench.cpp:119)
==131665==    by 0x1F3CB6: main (bench_bitcoin.cpp:132)
==131665==
==131665== Invalid read of size 8
==131665==    at 0x607B86: wallet::CWallet::AddToWalletIfInvolvingMe(std::shared_ptr<CTransaction const> const&, std::variant<wallet::TxStateConfirmed, wallet::TxStateInMempool, wallet::TxStateInactive> const&, bool, bool) (wallet.cpp:1252)==131665==    by 0x60879A: SyncTransaction (wallet.cpp:1409)
==131665==    by 0x60879A: wallet::CWallet::blockConnected(interfaces::BlockInfo const&) (wallet.cpp:1479)
==131665==    by 0x28B59C: generateFakeBlock(CChainParams const&, node::NodeContext const&, wallet::CWallet&, CScript const&) (wallet_create_tx.cpp:73)
==131665==    by 0x28C982: WalletCreateTx(ankerl::nanobench::Bench&, OutputType, bool, std::optional<PreSelectInputs>) [clone .constprop.0] [clone .isra.0] (wallet_create_tx.cpp:103)
==131665==    by 0x215E9A: operator() (std_function.h:590)
==131665==    by 0x215E9A: benchmark::BenchRunner::RunAll(benchmark::Args const&) (bench.cpp:119)
==131665==    by 0x1F3CB6: main (bench_bitcoin.cpp:132)
==131665==  Address 0x75707428686b70c7 is not stack'd, malloc'd or (recently) free'd
==131665==
==131665==
==131665== Process terminating with default action of signal 11 (SIGSEGV)
==131665==  General Protection Fault
==131665==    at 0x607B86: wallet::CWallet::AddToWalletIfInvolvingMe(std::shared_ptr<CTransaction const> const&, std::variant<wallet::TxStateConfirmed, wallet::TxStateInMempool, wallet::TxStateInactive> const&, bool, bool) (wallet.cpp:1252)==131665==    by 0x60879A: SyncTransaction (wallet.cpp:1409)
==131665==    by 0x60879A: wallet::CWallet::blockConnected(interfaces::BlockInfo const&) (wallet.cpp:1479)
==131665==    by 0x28B59C: generateFakeBlock(CChainParams const&, node::NodeContext const&, wallet::CWallet&, CScript const&) (wallet_create_tx.cpp:73)
==131665==    by 0x28C982: WalletCreateTx(ankerl::nanobench::Bench&, OutputType, bool, std::optional<PreSelectInputs>) [clone .constprop.0] [clone .isra.0] (wallet_create_tx.cpp:103)
==131665==    by 0x215E9A: operator() (std_function.h:590)
==131665==    by 0x215E9A: benchmark::BenchRunner::RunAll(benchmark::Args const&) (bench.cpp:119)
==131665==    by 0x1F3CB6: main (bench_bitcoin.cpp:132)
==131665==
==131665== HEAP SUMMARY:
==131665==     in use at exit: 43,227,660 bytes in 59,138 blocks
==131665==   total heap usage: 342,235 allocs, 283,097 frees, 77,605,937 bytes allocated
==131665==
==131665== LEAK SUMMARY:
==131665==    definitely lost: 0 bytes in 0 blocks
==131665==    indirectly lost: 0 bytes in 0 blocks
==131665==      possibly lost: 912 bytes in 3 blocks
==131665==    still reachable: 43,226,748 bytes in 59,135 blocks
==131665==         suppressed: 0 bytes in 0 blocks
==131665== Rerun with --leak-check=full to see details of leaked memory
==131665==
==131665== For lists of detected and suppressed errors, rerun with: -s
==131665== ERROR SUMMARY: 555 errors from 2 contexts (suppressed: 0 from 0)
Segmentation fault
```

achow101 commented:
Benchmark crash should be fixed.

josibake commented:
Concept ACK

achow101 commented:
Rebased.

josibake commented on src/wallet/wallet.cpp:
```diff
@@ -3897,11 +3904,16 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
         if (ExtractDestination(script, dest)) not_migrated_dests.emplace(dest);
     }
 
+    m_cached_spks.clear();
```
in "wallet: Use scriptPubKey cache in IsMine" (https://github.com/bitcoin/bitcoin/pull/26008/commits/3279e0e8078b92bc590d1ab65ba35f8d197edc12):

Why are we clearing the cache? Wouldn't this invalidate the cache for any existing descriptors I have? If I'm not supposed to do wallet migration in a wallet that already has descriptors, maybe it would be better to check that the cache is empty and throw an error here if not?

achow101 replied:
It was necessary since the cache was being refilled by `LegacyScriptPubKeyMan::MigrateToDescriptor` when it creates the `DescriptorScriptPubKeyMan` replacements. But since those are now made with a no-op callback, this is unnecessary, so I've removed this and replaced it with an `Assume(m_cached_spks.empty())`.

josibake commented on src/wallet/wallet.cpp:
```diff
@@ -1571,11 +1571,18 @@ isminetype CWallet::IsMine(const CTxDestination& dest) const
+    // Search the cache
+    const auto& it = m_cached_spks.find(script);
+    if (it != m_cached_spks.end()) {
+        // The cache is only used for descriptor wallets. If a script is in the cache, then it is automatically ISMINE_SPENDABLE.
+        return ISMINE_SPENDABLE;
```
in "wallet: Use scriptPubKey cache in IsMine" (https://github.com/bitcoin/bitcoin/pull/26008/commits/3279e0e8078b92bc590d1ab65ba35f8d197edc12):

While this assumption is true today (and probably always will be), it feels like we are having CWallet make a decision that should be left up to the SPKMan. What if we searched the cache the same way as before and returned the max result from the SPKMan(s)?

```suggestion
        isminetype result = ISMINE_NO;
        for (const auto& spk : it->second) {
            result = std::max(result, spk->IsMine(script));
        }
        return result;
    }
```

I didn't compile this and test the benchmark, but I doubt it hurts the performance gains much (if any).

achow101 replied:
Done

josibake commented on src/wallet/wallet.cpp:
```diff
@@ -3432,12 +3432,16 @@ ScriptPubKeyMan* CWallet::GetScriptPubKeyMan(const OutputType& type, bool intern
-            spk_mans.insert(spk_man_pair.second.get());
-        }
+
+    // Search the cache first
```
in "wallet: Use scriptPubKey cache in GetScriptPubKeyMans" (https://github.com/bitcoin/bitcoin/pull/26008/commits/a78ad3f40678d553cbe6a28ed9daaddd3097d196):

nit:
```suggestion
    // Search the cache for descriptor scriptPubKeyMans

achow101 replied:
The cache is intended to be somewhat generic, so if we had another type of ScriptPubKeyMan in the future, it would still be used, so leaving this comment as is.

josibake commented:
> Dropped the signing changes as they were causing issues with being able to sign transactions that we have the keys for but not the descriptors. https://github.com/bitcoin/bitcoin/pull/23417 would allow us to resolve the performance issues for signing while retaining this behavior.

Is this worth revisiting since your re-write of the cache? The signing stuff could be a follow-up PR, but if it's simple enough it would be really nice to include it here. Otherwise, someone using migrate wallet could still hit a wall even with the cache when trying to sign a transaction.

I started playing with the idea here: https://github.com/josibake/bitcoin/commit/5c3a2dad385ee5ce971ed9f3cfc35f9dfbadd0b8 but didn't spend very much time thinking about it, so feel free to ignore if its not helpful or if I'm missing something obvious.

achow101 commented:
> Might be worth revisiting this now that the cache is much simpler. The signing stuff could be a follow-up PR, but if it's simple enough it would be really nice to include it here. Otherwise, someone using migrate wallet could still hit a wall even with the cache when trying to sign a transaction.

I don't think it's possible to do caching for signing in a sane way that doesn't involve rewriting how we manage keys. The crux of the issue is that we will sign for things even if the wallet has only the keys, but not the scripts.

achow101 committed:
bench: Add a benchmark for ismine

achow101 committed:
wallet: Introduce a callback called after TopUp completes

After TopUp completes, the wallet containing each SPKM will want to know
what new scriptPubKeys were generated. In order for all TopUp calls
(including ones internal the the SPKM), we use a callback function in
the WalletStorage interface.

achow101 committed:
wallet: Cache scriptPubKeys for all DescriptorSPKMs

Have CWallet maintain a cache of all known scriptPubKeys for its
DescriptorSPKMs in order to improve performance of the functions that
require searching for scriptPubKeys.

achow101 committed:
wallet: Use scriptPubKey cache in IsMine

achow101 committed:
wallet: Use scriptPubKey cache in GetScriptPubKeyMans

achow101 committed:
wallet: Use scriptPubKeyCache in GetSolvingProvider

achow101 commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -374,7 +377,7 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
     bool TopUpChain(WalletBatch& batch, CHDChain& chain, unsigned int size);
 public:
-    LegacyScriptPubKeyMan(WalletStorage& storage, int64_t keypool_size) : ScriptPubKeyMan(storage), m_keypool_size(keypool_size) {}
+    LegacyScriptPubKeyMan(WalletStorage& storage, std::function<void(const std::set<CScript>&, ScriptPubKeyMan*)> topup_callback, int64_t keypool_size) : ScriptPubKeyMan(storage, topup_callback), m_keypool_size(keypool_size) {}
```
It was being used for migration as `LegacyScriptPubKeyMan::MigrateToDescriptor()` creates `DescriptorScriptPubKeyMan` which needs it. However, these can also just take a no-op function, and I've done that as well as your suggestion.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/commit/46b3c569cfd01a45e5d9cf9428a07b06652e3df9

Looks good, thanks for taking the suggestions! Having the callbacks seems like a really nice feature: based on one of your review comments, I took a stab at using the callbacks with fast wallet rescans and using them simplifies that code quite a bit.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3542,6 +3542,12 @@ void CWallet::LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc)
         AddScriptPubKeyMan(id, std::move(spk_manager));
     }
+
+    // Cache scriptPubKeys
+    ScriptPubKeyMan* spkm = m_spk_managers.at(id).get();
+    for (const auto& script : spkm->GetScriptPubKeys()) {
+        m_cached_spks[script].insert(spkm);
+    }
```
This does not seems to be necessary. At this point (`LoadDescriptorScriptPubKeyMan`), the spkm should't contain any script.

achow101 replied:
Good catch, removed.

furszy reviewed (commented):
Concept ACK

I have been considering the idea of using a probabilistic data structure instead of the deterministic one for the membership-test over the past few days. Mainly, because I'm a bit concerned about the unbounded cache size. I was going to share it today, but it requires more work and the local benchmarks show a substantial speed difference in favor of this approach.

So.. as feature freeze is close, will finish review in few hours. I believe this can be implemented simpler, and shorter, using a signal instead of providing the topup callback to all spkm constructors.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3740,7 +3769,7 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
         WalletLogPrintf("Update existing descriptor: %s\n", desc.descriptor->ToString());
         spk_man->UpdateWalletDescriptor(desc);
```
Removing the spks from the wallet's cache for a particular spkm requires iterating all of the spks and looking for ones that point to this spkm. I think that would be quite slow, and also redundant, since `AddWalletDescriptor` can only ever add scripts to the wallet. Arguably, `UpdateWalletDescriptor` shouldn't be clearing `m_map_script_pub_keys` either.

josibake commented:
reACK https://github.com/bitcoin/bitcoin/pull/26008/commits/a96647308ea40028806e2db6d35fb5dd14bea15f

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3891,11 +3902,16 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
         if (ExtractDestination(script, dest)) not_migrated_dests.emplace(dest);
     }
 
+    Assume(m_cached_spks.empty());
+
     for (auto& desc_spkm : data.desc_spkms) {
         if (m_spk_managers.count(desc_spkm->GetID()) > 0) {
             error = _("Error: Duplicate descriptors created during migration. Your wallet may be corrupted.");
             return false;
         }
+        for (const auto& script : desc_spkm->GetScriptPubKeys()) {
+            m_cached_spks[script].insert(desc_spkm.get());
+        }
```
nit: sounds like these two changes should be in the previous commit (cde66ceae).

achow101 replied:
Done

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3465,11 +3465,19 @@ std::unique_ptr<SigningProvider> CWallet::GetSolvingProvider(const CScript& scri
-        if (spk_man_pair.second->CanProvide(script, sigdata)) {
-            return spk_man_pair.second->GetSolvingProvider(script);
+    // Search the cache first
+    const auto& it = m_cached_spks.find(script);
+    if (it != m_cached_spks.end()) {
+        for (const auto& spkm : it->second) {
+            if (spkm->CanProvide(script, sigdata)) {
```
What is the difference between this (a966473) and the previous commit (f0c844bda) in terms of `CanProvide` usage?

In the previous commit, `CanProvide` usage was removed, while here you kept it.
I understand that you removed the `CanProvide` call from the previous commit because it is redundant (it calls IsMine internally which is exactly what `m_cached_spks` does) but I also agree that we might want to keep `CanProvide` because its internal implementation could vary over time.

achow101 replied:
I think this may have just been a holdover from when I was investigating using the same cache for signing. There should be no difference as `CanProvide` is redundant in both contexts.

achow101 replied:
I've dropped this `CanProvide`.

furszy reviewed (commented):
Left two more comments. They aren't really blocking but I'm still struggling with the memory consumption topic. Isn't `std::unordered_map` going to store all keys (entire scripts) in memory for duplicated so it can solve hash collisions?

Apart from that, just as an extra note for the future: It seems that we will need to rearrange code so that these functions aren't called so many times within the same process.

achow101 commented:
> They aren't really blocking but I'm still struggling with the memory consumption topic. Isn't `std::unordered_map` going to store all keys (entire scripts) in memory for duplicated so it can solve hash collisions?

Yes.

I've done some memory usage profiling with a large wallet, and various combinations of ordered and unordered maps and sets.

The baseline memory usage of `m_map_script_pub_keys`, which is the same for all of the following, is 1.2 MiB

* `std::unordered_map<CScript, std::unorderd_set<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks`: 3.4 MiB
* `std::unordered_map<CScript, std::set<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks`: 2.6 MiB
* `std::unordered_map<CScript, std::vector<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks`: 1.2 MiB
* `std::map<CScript, std::unorderd_set<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks`: 3.7 MiB
* `std::map<CScript, std::set<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks`: 2.9 MiB
* `std::map<CScript, std::vector<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks`: 1.5 MiB

In the latest push, I've changed to using `std::unordered_map<CScript, std::vector<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks` so the memory consumption should be about double the size of all scriptPubKeys.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1571,11 +1571,22 @@ isminetype CWallet::IsMine(const CTxDestination& dest) const
-    for (const auto& spk_man_pair : m_spk_managers) {
-        result = std::max(result, spk_man_pair.second->IsMine(script));
+
+    // Search the cache
```
In commit "wallet: Use scriptPubKey cache in IsMine" (523010b7bded819fdc11396b307d9772ec8ffe33)

Would be helpful to say what cache lookups are an alternative to in this context. Maybe add comment "Search m_cached_spks cache to only call IsMine on relevant SPKMs instead of calling it on everything in m_spk_managers." Could add similar comments to GetScriptPubKeyMan and GetSolvingProvider in following commits too.

achow101 replied:
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3436,12 +3436,16 @@ ScriptPubKeyMan* CWallet::GetScriptPubKeyMan(const OutputType& type, bool intern
     std::set<ScriptPubKeyMan*> spk_mans;
-    SignatureData sigdata;
-    for (const auto& spk_man_pair : m_spk_managers) {
-        if (spk_man_pair.second->CanProvide(script, sigdata)) {
```
In commit "wallet: Use scriptPubKey cache in GetScriptPubKeyMans" (38cfcf3435a5113606c8066833d49a4ab560aea9)

New code is no longer calling CanProvide. Would be good to use Assert or Assume to document that it should be true and verify integrity of the cache. (Same comment applies to GetSolvingProvider in next commit too)

achow101 replied:
Added `Assume`s.

ryanofsky reviewed (approved):
Code review ACK b20d882fd53c21098eb7858b2dbca84eafd2b344. I don't have much insight into performance impact of this change, but it seems like it should make migrated legacy wallets much faster (10 times faster loading in one case: https://github.com/bitcoin/bitcoin/pull/26008#issuecomment-1943469180) even if it does increase memory usage, and it seems to be implemented correctly. 

re: https://github.com/bitcoin/bitcoin/pull/26008#pullrequestreview-1876002633

> Apart from that, just as an extra note for the future: It seems that we will need to rearrange code so that these functions aren't called so many times within the same process.

I'm not sure what this is referring to, so would be curious

achow101 committed:
wallet: Retrieve ID from loaded DescSPKM directly

Instead of iterating m_spk_managers a DescriptorSPKM has been loaded in
order to get it's ID to compare, have LoadDescriptorSPKM return a
reference to the loaded DescriptorSPKM so it can be queried directly.

achow101 commented on src/wallet/external_signer_scriptpubkeyman.h:
```diff
@@ -13,11 +13,11 @@ namespace wallet {
   public:
-  ExternalSignerScriptPubKeyMan(WalletStorage& storage, WalletDescriptor& descriptor, int64_t keypool_size)
-      :   DescriptorScriptPubKeyMan(storage, descriptor, keypool_size)
+  ExternalSignerScriptPubKeyMan(WalletStorage& storage, std::function<void(const std::set<CScript>&, ScriptPubKeyMan*)> topup_callback, WalletDescriptor& descriptor, int64_t keypool_size)
```
Good point, I've changed this to use `WalletStorage` instead. This greatly reduces the diff.

achow101 commented:
I've also added one more commit that speeds up loading and migration as I noticed one spot where `m_spk_managers` was being iterated again during loading.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/26008/commits/ac246f68299d0dc208ae513dfad1a8fc91b5e6d4

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3436,12 +3436,18 @@ ScriptPubKeyMan* CWallet::GetScriptPubKeyMan(const OutputType& type, bool intern
+    Assume(std::all_of(spk_mans.begin(), spk_mans.end(), [&script, &sigdata](ScriptPubKeyMan* spkm) { return spkm->CanProvide(script, sigdata); }));
+
+    // Legacy wallet
+    if (IsLegacy()) spk_mans.insert(GetLegacyScriptPubKeyMan());
```
In commit "wallet: Use scriptPubKey cache in GetScriptPubKeyMans" (b9dd0fecda98bd14d8d305a69e6ce65f1e0a73fa)

This also needs to check GetLegacyScriptPubKeyMan()->CanProvide() to match previous behavior, I think

achow101 replied:
I think it's fine as `LegacyScriptPubKeyMan` is designed to be able to handle unknown scripts gracefully. But I've added a `CanProvide` check just for consistency.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3556,15 +3557,15 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
-        AddScriptPubKeyMan(id, std::move(spk_manager));
+        spk_manager = std::unique_ptr<ScriptPubKeyMan>(new DescriptorScriptPubKeyMan(*this, desc, m_keypool_size));
     }
+    return dynamic_cast<DescriptorScriptPubKeyMan&>(AddScriptPubKeyMan(id, std::move(spk_manager)));
```
In commit "wallet: Retrieve ID from loaded DescSPKM directly" (ac246f68299d0dc208ae513dfad1a8fc91b5e6d4)

Not important, but you could avoid the dynamic_cast and remove some boilerplate with:

```c++
DescriptorScriptPubKeyMan& CWallet::LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc)
{
    DescriptorScriptPubKeyMan* spkm;
    if (IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
        spkm = new ExternalSignerScriptPubKeyMan(*this, desc, m_keypool_size);
    } else {
        spkm = new DescriptorScriptPubKeyMan(*this, desc, m_keypool_size);
    }
    AddScriptPubKeyMan(id, std::unique_ptr<ScriptPubKeyMan>{spkm});
    return *spkm;
}
```

achow101 replied:
Done as suggested

ryanofsky reviewed (commented):
Code review ac246f68299d0dc208ae513dfad1a8fc91b5e6d4

Looks good, but wondering about a possible bug in the GetScriptPubKeyMans legacy case (see below)

furszy commented on src/bench/wallet_ismine.cpp:
```diff
@@ -0,0 +1,75 @@
+static void WalletIsMine(benchmark::Bench& bench, bool legacy_wallet, int num_combo = 0)
+{
+    const auto test_setup = MakeNoLogFileContext<TestingSetup>();
+    test_setup->m_args.ForceSetArg("-unsafesqlitesync", "1");
```
In 65c7997221:

This is not needed. The benchmark uses a db mock.

achow101 replied:
Removed

furszy commented on src/wallet/wallet.h:
```diff
@@ -994,7 +994,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     //! Instantiate a descriptor ScriptPubKeyMan from the WalletDescriptor and load it
-    void LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc);
+    DescriptorScriptPubKeyMan& LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc);
```
In bf898af0190d:

It was removed on the last push: `LoadDescriptorScriptPubKeyMan` now returns the spkm but no function utilizes it.

achow101 replied:
Oops, somehow that usage got dropped. Fixed.

furszy reviewed (commented):
Reviewed. Need to update the second commit description.

achow101 commented:
> Need to update the second commit description.

Done

ryanofsky reviewed (approved):
Code review ACK e041ed9b755468d205ed48b29f6c4b9e9df8bc9f. Just suggested changes since last review

josibake reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/26008/commits/e041ed9b755468d205ed48b29f6c4b9e9df8bc9f

Reviewed diff, looks good!

furszy reviewed (commented):
Code review ACK e041ed9b

It isn't blocking, but I have to admit that I'm still not really happy with the doubled script storage. I think we can do better. Will be experimenting with possible solutions.
A first measure to decrease it and remain fast, without changing the structure, could be to use the cache only for the inactive SPKMs. Since the active ones are limited in number (max 8), they can be checked quickly. However, this approach does not address the issue of handling really large scripts, which could probably be resolved by changing the structure for a probabilistic one.
